### ISSUE TITLE
Julia v1.11.2 pre-release: Package extension loading issue

### ISSUE THREAD

**Issue by mtfishman** [JuliaLang/julia#56656]
There appears to be an issue loading the LazyArraysBlockArraysExt package extension defined in [LazyArrays.jl](https://github.com/JuliaArrays/LazyArrays.jl) in Julia v1.11.1 and the [v1.11.2 pre-release](https://github.com/JuliaLang/julia/pull/56228), which doesn't appear in Julia v1.11.0 or Julia nightly.

Under certain conditions, when LazyArraysBlockArraysExt is being loaded in Julia v1.11.1 and the v1.11.2 pre-release, it fails with the error:
```julia
ERROR: LoadError: ArgumentError: Package LazyArraysBlockArraysExt does not have BlockArrays in its dependencies
```
as demonstrated in the CI logs of this PR: https://github.com/JuliaArrays/BlockArrays.jl/pull/433. 

I say "under certain conditions" because I'm only able to see the issue when it is run in Github Actions with the workflow defined in the BlockArrays repository, I haven't been able reproduce the issue locally (it must be related to the particular combination of packages that are loaded in the workflow). It might be related to #56426, #56204, #55939, but I'm raising a new issue since it isn't obvious to me if it is a new issue or a repeat of one of those. It looks most similar to #55939, however I removed BlockArrays from the dependency list of LazyArrays here: https://github.com/JuliaArrays/LazyArrays.jl/pull/347 and even after that the issue still shows up. We were hoping that #56228 would fix this issue but it doesn't seem like it has, since we tried testing against `1.11-nightly` in the workflow and see the same failure that we get when testing against `1.11.1` (but when we test against `1.11.0` or `nightly` the issue doesn't appear).

**Comment by topolarity** [comment]
I think the fix that's missing here is to backport this revert: https://github.com/JuliaLang/julia/pull/56234

**Comment by topolarity** [comment]
I was able to reproduce this locally by doing:
```julia
$ cd /path/to/BlockArrays.jl
$ rm -rf downstream depot
$ git clone https://github.com/JuliaArrays/LazyArrays.jl downstream
$ JULIA_DEPOT_PATH="$(pwd)/depot:" julia +pr56228 -q --project=downstream
julia> using Pkg

julia> Pkg.develop(PackageSpec(path="."))  # resolver may fail with main deps
   Resolving package versions...
    Updating `~/repos/BlockArrays.jl/downstream/Project.toml`
  [4c555306] + ArrayLayouts v1.10.4
  [8e7c35d0] + BlockArrays v1.2.0 `..`
  [1a297f60] + FillArrays v1.13.0
  [1914dd2f] + MacroTools v0.5.13
  [37e2e46d] ~ LinearAlgebra ⇒ v1.11.0
  [2f01184e] ~ SparseArrays ⇒ v1.11.0
    Updating `~/repos/BlockArrays.jl/downstream/Manifest.toml`
# ...

julia> Pkg.test() # triggers a failing pre-compile (Pkg.precompile is fine, for some reason)
```

I'm not 100% sure since repro was a little spotty, but it seems like https://github.com/JuliaLang/julia/pull/56658 indeed resolves the problem

**Comment by topolarity** [comment]
This should now be fixed in the recently released 1.11.2

Please reopen if you see any lingering issues.
